### DISCUSSION TITLE
Number of steps per epoch

### DISCUSSION THREAD

**Discussion by celsofranssa** [Lightning-AI/pytorch-lightning#5444] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule]
Some learning rate schedulers as [OneCycleLR](https://pytorch.org/docs/stable/optim.html#torch.optim.lr_scheduler.OneCycleLR) requires the number of steps per epoch.
Then, how to get the number of steps in `configure_optimizers(self)` scope?

**Comment by teddykoker** [comment]
Normally I pass the total number of samples in the dataset as well as the batch size into my model's `__init__()`. Then the number of steps per epoch can be computed with `self.samples // self.batch_size`.

**Reply by celsofranssa** [reply]
Currently, I am using the following (for single device):

```python
    def num_training_steps(self) -> int:
        """Total training steps inferred from datamodule and number of epochs."""
        steps_per_epochs = len(self.train_dataloader()) / self.trainer.accumulate_grad_batches
        max_epochs = self.trainer.max_epochs
        return steps_per_epochs * max_epochs
``` 
However, maybe exist a better solution as pointed out in #5449.
